scheduler/scheduler.go

```diff
@@ -220,10 +285,101 @@ func spawnHosts(ctx context.Context, newHostsNeeded map[string]int) (map[string]
 		"span":      time.Since(startTime).String(),
 		"duration":  time.Since(startTime),
 	})
-
 	return hostsSpawnedPerDistro, nil
 }
 
+// IsMaxHostsRunning returns true if the max number of hosts are already running
+func IsMaxHostsRunning(d distro.Distro, allDistroHosts []host.Host) bool {
+	if d.MaxContainers != 0 {
+		if len(allDistroHosts) >= (d.PoolSize * d.MaxContainers) {
+			grip.Info(message.Fields{
+				"distro":         d.Id,
+				"runner":         RunnerName,
+				"pool_size":      d.PoolSize,
+				"max_containers": d.MaxContainers,
+				"message":        "max hosts running on all containers",
+			})
+			return true
+		}
+	}
+
+	if len(allDistroHosts) >= d.PoolSize {
+		grip.Info(message.Fields{
+			"distro":    d.Id,
+			"runner":    RunnerName,
+			"pool_size": d.PoolSize,
+			"message":   "max hosts running",
+		})
+		return true
+	}
+
+	return false
+}
+
+// GenerateHostOptions generates host options based on what kind of host it is:
+// regular host or container
+func GenerateHostOptions(d distro.Distro) (cloud.HostOptions, error) {
+	if d.MaxContainers != 0 {
+		parent, err := FindAvailableParent(d)
+		if err != nil {
+			return cloud.HostOptions{}, err
+		}
+		log.Print("id ", parent.Id)
+		hostOptions := cloud.HostOptions{
+			ParentID: parent.Id,
+			UserName: evergreen.User,
+			UserHost: false,
+		}
+		return hostOptions, nil
+	}
+
+	hostOptions := cloud.HostOptions{
+		UserName: evergreen.User,
+		UserHost: false,
+	}
+	return hostOptions, nil
+}
+
+// GenerateParentHostOptions generates host options for a parent host
+func GenerateParentHostOptions() cloud.HostOptions {
+	return cloud.HostOptions{
+		HasContainers: true,
+		UserName:      evergreen.User,
+		UserHost:      false,
+	}
+}
+
+// FindAvailableParent finds a parent host that can accommodate container
+// based on oldest parent
+func FindAvailableParent(d distro.Distro) (host.Host, error) {
+	AllParents, err := host.FindAllRunningParents()
+	if err != nil {
+		return host.Host{}, err
+	}
+
+	OldestParent := host.Host{CreationTime: time.Now()}
+	for _, parent := range AllParents {
```

Comment:
If there are no running parents, then this function will return a pseudo-host with just creation time set. It might be better to keep track of the oldest parent creation time without using a host document.